Integration tests

paweljakubas opened the pull request:
There are two suites of tests to be added.
1. Instantiating websocket server with client and showing that websocket server emulating hydra emits the expected events. Running:
```
cargo test hydra_emulation -- --nocapture
```
2. Using the websocket server emulating hydra and treating it as source and connecting to hydra-oura-stdout pipeline as driven by examples/hydra.config.toml. This running:
```
cargo test hydra_oura_stdout -- --nocapture
```
Here we use dedicated daemon.toml which produces json. Every run it is truncated to 0, but one can inspect how it looks like -> `tests/hydra/logs.txt`

paweljakubas committed:
add tokio runtime skeleton in integration testing

paweljakubas committed:
websocket server-client scaffolding

paweljakubas committed:
towards hydra emulation test

paweljakubas committed:
finish hydra emulation testing

paweljakubas committed:
try running binary

paweljakubas committed:
add oura integration test working

paweljakubas committed:
add scenario 2

refining testing plus gitignore

spawn cmd process and fix typo

truncate log file

make sure result of tests is taken into account

clean redundant tests

Anviking commented on tests/hydra.rs:
```diff
@@ -440,3 +452,209 @@ fn scenario_2() -> TestResult {
+}
+
+#[test]
+fn hydra_oura_stdout_scenario_1() -> TestResult {
+    let tx_event = r#"{"event":"apply","point":{"hash":"84e657e3dd5241caac75b749195f78684023583736cc08b2896290ab","slot":7},"record":{"hex":"84a300d9010281825820f0a39560ea80ccc68e8dffb6a4a077c8927811f06c5d9058d0fa2d1a8d047d2000018282581d605e4e214a6addd337126b3a61faad5dfe1e4f14f637a8969e3a05eefd1a001e848082581d600d45f2b310a98e766cee2ab2f6756c91719bd7b35929cef058365b651a015ef3c00200a100d90102818258 ...
+    let events = vec![
+        "PeerConnected".to_string(),
+        "PeerConnected".to_string(),
+        "Greetings".to_string(),
+        "HeadIsInitializing".to_string(),
+        "Committed".to_string(),
+        "Committed".to_string(),
+        "Committed".to_string(),
+        "HeadIsOpen".to_string(),
+        "TxValid".to_string(),
+        tx_event.to_string(),
+        "SnapshotConfirmed".to_string(),
+        "HeadIsClosed".to_string(),
+        "ReadyToFanout".to_string(),
+        "HeadIsFinalized".to_string(),
+    ];
+    hydra_oura_stdout_test("tests/hydra/scenario_1.txt".to_string(), events)
```
I tried mutating the expectations: tweaking 1) tx binary, 2) tx event slot, 3) dropping an event, but the test appears to still pass

paweljakubas replied:
now it signals when failing. nice catch.thanks

Anviking commented on tests/hydra.rs:
```diff
@@ -440,3 +452,209 @@ fn scenario_2() -> TestResult {
+
+#[test]
+fn hydra_emulation_scenario_2() -> TestResult {
+    hydra_emulation_test("tests/hydra/scenario_2.txt".to_string())
```
I don't understand, what is the purpose of the two `hydra_emulation_test` tests which have a clustom websocket client as opposed to testing `oura`? If this is a sanity-check from development it shouldn't be needed any longer

paweljakubas replied:
removed those redundant tests - although were of use to me when I built the final one as an intermediate steps

Anviking commented on tests/hydra.rs:
```diff
@@ -440,3 +452,209 @@ fn scenario_2() -> TestResult {
+        predicates.push(predicate_fn.eval(json));
+        count += 1;
+    }
+    assert_eq!(predicates, vec![true; expected.len()]);
```
Off the top of my head, I find it odd that the `oura_pipeline` is responsible for both capturing the output and performing the assertion. Did you consider returning the captured outputs in the result? 

And — maybe we could check the output with something like [goldenfile](https://crates.io/crates/goldenfile) instead of `assert_eq` if the output is deterministic? 🤔

paweljakubas replied:
added golden as suggested :+1:

paweljakubas committed:
add golden

Anviking committed:
factor out `run_daemon`

Anviking committed:
factor out `mock_hydra_node`

Anviking committed:
Tweak integration tests

- Use random TCP ports
- Use random output files
- Use goldenfile (has the practical UPDATE_GOLDENFILES=1 flag)

Anviking reviewed (commented):
Wonderful to have 👍 I created a PR with some tweaks to this PR https://github.com/Anviking/oura/pull/10

paweljakubas committed:
Merge pull request #10 from Anviking/anviking/integration-test-tweaks

Tweak integration tests
